scrapy/utils/asyncgen.py

```diff
@@ -0,0 +1,31 @@
+"""
+Helpers using Python 3.6+ async generator syntax (ignore SyntaxError on import).
+"""
+import collections
+
+
+async def collect_asyncgen(result):
+    results = []
+    async for x in result:
+        results.append(x)
+    return results
+
+
+async def as_async_generator(it):
+    if isinstance(it, collections.abc.AsyncIterator):
+        async for r in it:
+            yield r
+    else:
+        for r in it:
+            yield r
+
+
+async def process_async_iterable_helper(it, in_predicate=None, out_predicate=None, processor=None):
```

Comment:
This is my idea of a generic helper for some middleware methods that process an iterable. I'm sure there can be cleaner things, so I'm open for suggestions. I've looked at our spider middlewares and several 3rd-party ones and it loks like most of them can be rewritten to use this, but the result is not always clean.

It would be much cleaner if we could just process every item of an iterable, but middlewares also often drop them (hence the predicate functions), which I think can only be done in a per-item processor function by raising a special exception which I don't like.

Reply:
an idea: could we have a helper which converts a standart generator to asnyc generator, so that users can write `async for` always? It looks pretty awkward to limit middlewares to predicates and processors.

Reply:
Yes, `as_async_generator` is a such helper.

Also, it's not limited, you just need to write your code twice. I wrote this helper to be able to convert all our spider middlewares (their `process_spider_output` methods) to support normal and async callbacks without code duplication.

Reply:
Hey! I'd prefer to remove process_async_iterable_helper, process_normal_iterable_helper and process_iterable_helper from the code here: they are unused in Scrapy itself, not really documented; it'd make the PR smaller.